### PR TITLE
[ROCm] Single-kernel gated delta net prefill for RDNA3.5

### PR DESCRIPTION
**Pull request by roberteg16**
## Summary

GATED_DELTA_NET prefill runs as four Triton kernels that communicate through HBM. The chunk recurrence hands its state to the output kernel as `h` of shape `[NT, H, V, K]` — 32 MiB stored and loaded back per layer at 2048 tokens — and none of it outlives the chunk that produced it.

This adds the chunked delta rule (WY representation / UT transform) as a single HIP kernel, ported from llama.cpp. One block walks the chunks of one (head, sequence) with the state resident in registers and the per-chunk intermediates in LDS, so `h`, `w` and `u` are never formed. A 32-token chunk and the identity `v_new = Tinv (V beta - (K beta exp(g_cs)) S)`, which avoids materialising `U` and `W`, are what keep a whole chunk inside the 64 KB LDS budget.

It is also more accurate than the path it replaces: everything between the bf16 inputs and the bf16 output stays fp32, where the Triton kernels round `h`, `w`, `u`, `Tinv` and `v_new` to bf16.

### Kernel by kernel

What the one kernel replaces, at the Qwen3.5/3.6 35B-A3B shape (32 value heads, 16 key heads, head dim 128). Each Triton kernel timed on its own, times in ms:

| kernel | 512 | 1024 | 2048 | 4096 | 8192 |
|---|---|---|---|---|---|
| `chunk_local_cumsum` | 0.007 | 0.009 | 0.011 | 0.020 | 0.039 |
| `kkt` + `solve_tril` + `wy` | 0.309 | 0.669 | 1.910 | 4.839 | 13.738 |
| `chunk_delta_h` | 0.295 | 0.495 | 0.854 | 1.545 | 2.956 |
| `chunk_o` | 0.476 | 1.026 | 2.406 | 6.133 | 14.920 |
| four kernels, isolated | 1.087 | 2.198 | 5.181 | 12.537 | 31.653 |
| **`gdn_chunked`, one kernel** | **0.649** | **1.256** | **2.413** | **4.743** | **9.535** |

The isolated figures each carry their own sync and cache flush, so below about 2048 tokens they over-count against the pipelined whole-op time below; at 8192 the two agree to under 1%.

### Whole op

Same run, caches flushed between iterations and measured at sustained clocks:

| tokens | 512 | 1024 | 2048 | 4096 | 8192 |
|---|---|---|---|---|---|
| triton | 0.765 | 1.877 | 4.848 | 12.322 | 31.431 ms |
| hip | **0.649** | **1.256** | **2.413** | **4.743** | **9.535** ms |
| | 1.18x | 1.49x | 2.01x | 2.60x | 3.30x |

Host-side dispatch drops from about 0.18 ms per call to 0.013 ms, four launches becoming one.

### End to end

Qwen3.6-35B-A3B-W4A16, 15368 in / 2000 out, 10 prompts, three paired runs per configuration alternating baseline and kernel. Run at `--max-num-seqs 1`, so every step has a single prefill sequence and the kernel is at its least parallel; the table below is a floor rather than a best case. Ranges do not overlap on any metric:

| | triton | hip | |
|---|---|---|---|
| TTFT | 8913 ms | **8441 ms** | -5.3% |
| prefill | 1824 tok/s | **1926 tok/s** | +5.6% |
| e2e (median) | 40721 ms | **40376 ms** | -0.9% |

Decode is unaffected — it uses the recurrent kernel, not this path.

### Short prompts

The gain scales with prefill length, so the short-prompt multimodal case is worth stating separately. Same model, 100 text tokens plus one synthetic image, 128 output tokens, 10 prompts, three paired runs per configuration with the order alternating within each pair.

| | prefill tokens | TTFT | prefill | e2e (median) |
|---|---|---|---|---|
| 640x480, triton | 412 | 252.3 ms | 1635 tok/s | 1869 ms |
| 640x480, hip | 412 | 247.0 ms | 1670 tok/s | 1864 ms |
| | | -2.1% | +2.2% | -0.3% |
| 1280x720, triton | 992 | 536.7 ms | 1849 tok/s | 2168 ms |
| 1280x720, hip | 992 | **522.3 ms** | **1899 tok/s** | **2153 ms** |
| | | **-2.7%** | **+2.7%** | **-0.7%** |

At 992 tokens the three-run ranges do not overlap on TTFT, prefill or e2e, so the improvement is real but small: the op is a few percent of TTFT at this length, and TTFT is itself a fraction of a 128-token request. At 412 tokens the ranges do overlap and the result is within run-to-run noise. Neither size regresses, and decode is unchanged in both at 78 tok/s.

The 1280x720 figure is consistent with the kernel measurement. At 1024 tokens the kernel saves 0.621 ms per call, the model has 30 gated delta net layers out of 40, so a prefill step should shed about 18.6 ms of a 537 ms TTFT, or 3.5%. The observed 14.4 ms is the same quantity within the share of TTFT that is not this op.

### Where it wins

Whole op against prefill length, one sequence, times in ms:

| tokens | chunks | triton | hip | ratio |
|---|---|---|---|---|
| 1 | 1 | 0.158 | 0.076 | **2.07x** |
| 32 | 1 | 0.177 | 0.086 | **2.05x** |
| 33 | 2 | 0.176 | 0.118 | **1.49x** |
| 64 | 2 | 0.192 | 0.125 | **1.54x** |
| 128 | 4 | 0.235 | 0.205 | **1.14x** |
| 192 | 6 | 0.297 | 0.283 | **1.05x** |
| 256 | 8 | 0.350 | 0.359 | 0.98x |
| 320 | 10 | 0.424 | 0.433 | 0.98x |
| 384 | 12 | 0.495 | 0.508 | 0.98x |
| 448 | 14 | 0.661 | 0.586 | **1.13x** |
| 512 | 16 | 0.766 | 0.664 | **1.15x** |
| 1024 | 32 | 1.874 | 1.259 | **1.49x** |

There is no lower bound: at one token the kernel is 2.07x, because the Triton path has a floor of about 0.16 ms from four eager launches against this path's one, and prefill is not CUDA-graph captured. The 32-token chunk shows up in the cost function rather than as a crossover — the step from 32 to 33 tokens is the second chunk being added (0.086 to 0.118 ms), and beyond that the kernel is linear at about 0.039 ms per chunk.

The one region where it loses is 256 to 384 tokens, by 2 to 4%. That is not a property of the length but of how few blocks a single sequence provides: the grid is (heads, sequences) and 40 blocks are resident at once, so one sequence fills 32 of them. Same total tokens, split across more requests:

| total tokens | 1 seq | 2 seqs | 4 seqs | 8 seqs |
|---|---|---|---|---|
| 256 | 0.98x | 0.98x | **1.08x** | **2.00x** |
| 320 | 0.96x | **1.06x** | **1.35x** | **1.45x** |
| 384 | 0.97x | **1.02x** | **1.45x** | **1.50x** |

The Triton path moves the other way, getting slower as the same tokens are spread over more requests. So the losing case needs both few tokens and a single sequence at once, where the op costs 0.4 ms in absolute terms. A gate would have to key on the pair to catch it, for at most 4%, so there is none.

### Scope and gating

Built on RDNA3.5. The block layout assumes wave32 throughout, and the device-side fallbacks would otherwise let it run and return wrong numbers on wave64, so it is gated three ways: the source is only compiled where it is valid, the op is only registered there, and the host function rechecks `gcnArchName` and `warpSize`. Head dim 128, varlen and bf16 only; everything else falls back to the Triton kernels, which are untouched. `VLLM_GDN_HIP=0` disables it.

The family is not uniform in register file size, and the launch bounds follow it. gfx1151 has 1536 registers per SIMD, so asking for 16 waves per SIMD caps the allocation at 96 and two blocks stay resident. The other three have 1024, where the same request caps at 64 and spills 32 registers per thread with no diagnostic, so they ask for 8 instead: 100 registers, no spill, one block per WGP.

Only gfx1151 has been run on hardware. The other three are verified at the compilation level — they build, they do not spill, and the occupancy is known — but no token has gone through them.

## Test plan

- [x] `pytest tests/kernels/mamba/test_gdn_chunked_rocm.py` — 21 cases pass: op registration, GQA ratios 1/2/4, partial chunks including `[32, 33, 64, 1]` and ragged batches, the all-single-token batch that speculative decoding reclassifies into this path, fp32 and bf16 initial state, `core_attn_out` aliasing against a NaN-poisoned buffer, the gate declining every case the kernel asserts on, and `torch.library.opcheck` over the schema and the registered fake.
- [x] `benchmarks/kernels/benchmark_gdn_chunk.py --check --exact` against an fp64 token-by-token reference: relative RMS error 0.53x the Triton path's across single, ragged and short-sequence batches.
- [x] End-to-end sanity on Qwen3.6-35B-A3B-W4A16 (multimodal check passes on both paths).
- [x] `pre-commit run` clean on all touched files.
- [x] Compiled for every RDNA3.5 target with `-Rpass-analysis=kernel-resource-usage`: LDS 58624 B throughout, no new register spills, and 16 waves/SIMD on gfx1151 against 9 on the smaller-register-file parts.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1354,6 +1354,17 @@ if(VLLM_GPU_LANG STREQUAL "HIP")
       "csrc/rocm/moe_q_gemm_rdna3.cu")
   endif()
 
+  # Single-kernel gated delta net prefill.  The block layout assumes wave32
+  # throughout, so the source is only built where it is valid; the host
+  # function rechecks gcnArchName at launch, because the device-side fallbacks
+  # would otherwise let it run and return wrong numbers on wave64.
+  set(VLLM_ROCM_HAS_GFX115X OFF)
+  if(VLLM_GPU_ARCHES MATCHES "gfx115[01]")
+    set(VLLM_ROCM_HAS_GFX115X ON)
```
Comment (eble-amd) [inline CMakeLists.txt]: Misleading name: a variable named `VLLM_ROCM_HAS_GFX115X` should be `ON` for any GPU matching `gfx115`, including `gfx1152` and `gfx1153`.  If this task calls for building a file for gfx1150 and gfx1151 only, a different name (maybe two variables) should be used.

File: vllm/model_executor/layers/fla/ops/chunk_rocm.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+import functools
+import os
+
+import torch
+
+from vllm.platforms.rocm import on_gfx1150, on_gfx1151
+
+# Set VLLM_GDN_HIP=0 to fall back to the Triton kernels.
+_ENABLED = os.getenv("VLLM_GDN_HIP", "1") == "1"
```
Comment (eble-amd) [inline vllm/model_executor/layers/fla/ops/chunk_rocm.py]: Should this be in vllm/envs.py?

File: csrc/rocm/gdn_chunked.cu
Comment (eble-amd) [inline csrc/rocm/gdn_chunked.cu:1]: Clod is not a lawyer, but it reckons that the MIT license from llama.cpp should be carried over with ported code.

File: vllm/model_executor/layers/fla/ops/chunk.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
 
 from .chunk_delta_h import chunk_gated_delta_rule_fwd_h
 from .chunk_o import chunk_fwd_o
+from .chunk_rocm import chunk_gdn_hip_fwd, is_hip_gdn_supported
 from .chunk_scaled_dot_kkt import chunk_scaled_dot_kkt_fwd
```
Comment (eble-amd) [inline vllm/model_executor/layers/fla/ops/chunk.py:20]: Does this import or the code below need to be guarded to run on non-ROCm platforms?  This is not a big concern for our branch, but will need to be understood before upstreaming.

File: csrc/rocm/gdn_chunked.cu [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+  return __builtin_amdgcn_exp2f(x);
+}
+
+__device__ __forceinline__ float gdn_bf16_to_f32(uint16_t x) {
+  union {
+    uint32_t u;
+    float f;
+  } c;
+  c.u = static_cast<uint32_t>(x) << 16;
+  return c.f;
```
Comment (eble-amd) [inline csrc/rocm/gdn_chunked.cu]: Writing one union member and reading another invokes undefined behavior in C++.  Instead:

```cpp
return std::bit_cast<float>(uint32_t{x} << 16);
```

It also happens in `gdn_f32_to_bf16()` immediately below.

And BTW, `__builtin_bit_cast` used elsewhere in this file could be changed to `std::bit_cast`.

File: csrc/rocm/gdn_chunked.cu [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+
+// Lanes cooperating on one adjacent pair of state columns.  Sixteen of them fit
+// inside a wave32, so their reductions stay cross-lane with no LDS traffic and
+// no barrier, and holding a pair rather than a single column lets every staged
+// K/Q element feed two FMAs.
+constexpr int GDN_COL_LANES = 16;
+constexpr int GDN_ROWS_PER_LANE = GDN_HEAD_DIM / GDN_COL_LANES;
+
+#if defined(__gfx1150__) || defined(__gfx1151__)
+  #define GDN_RDNA3_5 1
```
Comment (eble-amd) [inline csrc/rocm/gdn_chunked.cu]: Misleading name: gfx1152 and gfx1153 are also RDNA 3.5 GPUs, right?

**Review by eble-amd — COMMENTED** [review]
IMO, the undefined behavior should be taken care of before merging.  The other things could be deferred.

File: CMakeLists.txt (hunk 1 above)
Comment (roberteg16) [reply CMakeLists.txt]: done

File: vllm/model_executor/layers/fla/ops/chunk_rocm.py (hunk 2 above)
Comment (roberteg16) [reply vllm/model_executor/layers/fla/ops/chunk_rocm.py]: done

File: csrc/rocm/gdn_chunked.cu
Comment (roberteg16) [reply csrc/rocm/gdn_chunked.cu:1]: done

File: vllm/model_executor/layers/fla/ops/chunk.py (hunk 3 above)
Comment (roberteg16) [reply vllm/model_executor/layers/fla/ops/chunk.py:20]: done

File: csrc/rocm/gdn_chunked.cu (hunk 4 above)
Comment (roberteg16) [reply csrc/rocm/gdn_chunked.cu]: done

File: csrc/rocm/gdn_chunked.cu (hunk 5 above)
Comment (roberteg16) [reply csrc/rocm/gdn_chunked.cu]: done

Comment (roberteg16) [reply csrc/rocm/gdn_chunked.cu]: Yes, they are. The kernel now builds and is registered for all four: gfx1150, gfx1151, gfx1152 and gfx1153. The macro is `GDN_SUPPORTED_ARCH` and enumerates them, because there is no gfx11.5 family macro to use instead — the compiler gives you a per-target `__gfxNNNN__` and `__GFX11__`, and the latter also covers gfx1100 through gfx1103.

Widening it turned up a tuning bug that predates the PR. The family is not uniform in register file size: gfx1151 has 1536 registers per SIMD, the other three have 1024. `__launch_bounds__` asking for 16 waves per SIMD therefore caps the allocation at 96 on gfx1151, where two blocks stay resident, but at 64 on the rest, where it spills 32 registers per thread — silently, no diagnostic. gfx1150 was already shipping in that state. Those three now ask for 8 instead, which gives 100 registers and no spill, at one block per WGP.

Worth stating plainly: only gfx1151 has been run on hardware. The other three are verified at the compilation level (they build, do not spill, and the occupancy is known), but nobody has put a token through them.

File: CMakeLists.txt (hunk 1 above)
Comment (roberteg16) [reply CMakeLists.txt]: Renamed to `VLLM_ROCM_HAS_GDN_CHUNKED`, and the compile definition to `VLLM_ROCM_GDN_CHUNKED`, so the gate names the kernel rather than an arch family. The match is now `gfx115[0-3]`, since the kernel is enabled on all four of gfx1150, gfx1151, gfx1152 and gfx1153.

**Review by eble-amd — APPROVED** [review]
